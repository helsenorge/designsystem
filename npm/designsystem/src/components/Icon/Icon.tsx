import React from 'react';

import classNames from 'classnames';

import { AnalyticsId, IconSize } from '../../constants';
import { useUuid } from '../../hooks/useUuid';

export type SvgIcon = React.FC<SvgPathProps>;

export interface BaseIconProps {
  /* aria-label for the <svg> element. Used as <title> tag. */
  ariaLabel?: string;
  /* Changes the size of the icon. */
  size?: number;
  /* Changes the color of the icon. */
  color?: string;
  /* Changes the hover color of the icon. */
  hoverColor?: string;
  /* Adds custom classes to the element. */
  className?: string;
  /* Swaps the displayed icon to the hover version and changes its color. */
  isHovered?: boolean;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export interface IconProps extends BaseIconProps {
  /* Sets which icon should be displayed. */
  svgIcon: SvgIcon;
}

export interface SvgPathProps {
  size: IconSize;
  isHovered: boolean;
}

interface IconConfig {
  size: IconSize;
  isHovered: boolean;
  normal: React.ReactElement;
  normalHover: React.ReactElement;
  xSmall?: React.ReactElement;
  xSmallHover?: React.ReactElement;
  xxSmall?: React.ReactElement;
  xxSmallHover?: React.ReactElement;
}

export const getIcon = ({
  size,
  isHovered,
  normal,
  normalHover,
  xSmall,
  xSmallHover,
  xxSmall,
  xxSmallHover,
}: IconConfig): React.ReactElement => {
  if (size <= IconSize.XXSmall && xxSmall && xxSmallHover) {
    return isHovered ? xxSmallHover : xxSmall;
  }
  if (size <= IconSize.XSmall && xSmall && xSmallHover) {
    return isHovered ? xSmallHover : xSmall;
  }

  return isHovered ? normalHover : normal;
};

export const Icon = React.forwardRef((props: IconProps, ref: React.ForwardedRef<SVGSVGElement>) => {
  const {
    svgIcon,
    ariaLabel,
    className = '',
    size = IconSize.Small,
    color = 'black',
    hoverColor = color || 'black',
    isHovered = false,
    testId,
    ...other
  } = props;

  const svgRaw = React.createElement(svgIcon, {
    size,
    isHovered,
  });

  const titleId = useUuid();
  const svgColor = isHovered ? hoverColor : color;

  return (
    <svg
      data-testid={testId}
      data-analyticsid={AnalyticsId.Icon}
      ref={ref}
      className={classNames(`hnds-style-icon`, className)}
      role={ariaLabel ? 'img' : 'presentation'}
      aria-labelledby={ariaLabel ? titleId : undefined}
      focusable={false}
      aria-hidden={ariaLabel ? undefined : true}
      viewBox="0 0 48 48"
      style={{ minWidth: size, minHeight: size }}
      width={size}
      height={size}
      fill={svgColor}
      color={svgColor}
      {...other}
    >
      {ariaLabel && <title id={titleId}>{ariaLabel}</title>}
      {svgRaw}
    </svg>
  );
});

Icon.displayName = 'Icon';

export default Icon;
