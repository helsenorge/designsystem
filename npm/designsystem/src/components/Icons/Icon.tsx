import React from 'react';
import { AnalyticsId, IconSize } from '../../constants';

type SvgIcon = React.FC<SvgPathProps>;

interface IconProps {
  /* Sets which icon should be displayed. */
  svgIcon: SvgIcon;
  /* aria-label for the <svg> element. Used as <title> tag if id is set */
  ariaLabel?: string;
  /* id to */
  id?: string;
  /* Changes the size of the icon. */
  size?: number;
  /* Changes the color of the icon. */
  color?: string;
  /* Changes the hover color of the icon. */
  hoverColor?: string;
  /* Adds custom classes to the element. */
  className?: string;
  /* Use a custom hover icon. */
  hoverType?: SvgIcon;
  /* Swaps the displayed icon to the hover version and changes its color. */
  isHovered?: boolean;
  /** Sets the data-testid attribute. */
  testId?: string;
}

interface SvgPathProps {
  size: IconSize;
  isHovered: boolean;
}

export const getIcon = (
  size: IconSize,
  isHovered: boolean,
  normal: JSX.Element,
  normalHover: JSX.Element,
  small?: JSX.Element,
  smallHover?: JSX.Element,
  xxSmall?: JSX.Element,
  xxSmallHover?: JSX.Element
): JSX.Element => {
  if (size === IconSize.XSmall && small && smallHover) {
    return isHovered ? smallHover : small;
  }

  if (size === IconSize.XXSmall && xxSmall && xxSmallHover) {
    return isHovered ? xxSmallHover : xxSmall;
  }

  return isHovered ? normalHover : normal;
};

const Icon = React.forwardRef((props: IconProps, ref: React.ForwardedRef<SVGSVGElement>) => {
  const {
    svgIcon,
    ariaLabel,
    id,
    className = '',
    size = IconSize.Small,
    color = 'black',
    hoverColor = color || 'black',
    isHovered = false,
    testId,
    ...other
  } = props;

  const svgRaw = React.createElement(svgIcon, {
    size,
    isHovered,
  });

  const labelledby = id && ariaLabel ? `title-${id}` : undefined;

  return (
    <svg
      id={id}
      data-testid={testId}
      data-analyticsid={AnalyticsId.Icon}
      ref={ref as React.RefObject<SVGSVGElement>}
      className={`hnds-style-icon ${className}`}
      role="img"
      aria-label={!id && ariaLabel ? ariaLabel : undefined}
      aria-labelledby={labelledby}
      aria-hidden="true"
      viewBox="0 0 48 48"
      style={{ minWidth: size, minHeight: size }}
      width={size}
      height={size}
      fill={isHovered ? hoverColor : color}
      {...other}
    >
      {labelledby && <title id={labelledby}>{ariaLabel}</title>}
      {svgRaw}
    </svg>
  );
});

Icon.displayName = 'Icon';

export { Icon, IconProps, SvgPathProps, SvgIcon };
