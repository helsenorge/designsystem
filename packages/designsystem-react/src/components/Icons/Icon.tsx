import React from 'react';
import StyledIcon from './Icon.styled';

type SvgIcon = React.FC<SvgPathProps>;

interface IconProps {
  svgIcon: SvgIcon;
  ariaLabel?: string;
  size?: number;
  color?: string;
  hoverColor?: string;
  className?: string;
  hoverType?: SvgIcon;
  isHovered?: boolean;
  testId?: string;
}

interface SvgPathProps {
  isExtraSmall: boolean;
  isHovered: boolean;
}

const Icon = React.forwardRef((props: IconProps, ref: any) => {
  const {
    svgIcon,
    ariaLabel,
    className = '',
    size = 48,
    color = 'black',
    hoverColor = color || 'black',
    isHovered = false,
    testId,
    ...other
  } = props;

  const svgRaw = React.createElement(svgIcon, {
    isExtraSmall: size <= 38,
    isHovered,
  });

  return (
    <StyledIcon
      data-testid={testId}
      ref={ref}
      className={`hnds-style-icon ${className}`}
      role="img"
      aria-label={ariaLabel}
      aria-hidden="true"
      viewBox="0 0 48 48"
      width={size}
      height={size}
      fill={isHovered ? hoverColor : color}
      {...other}>
      {svgRaw}
    </StyledIcon>
  );
});

Icon.displayName = 'Icon';

export {Icon, IconProps, SvgPathProps, SvgIcon};
