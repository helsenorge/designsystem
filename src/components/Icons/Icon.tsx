import React from 'react';

import Lock from './Lock';
import Unlock from './Unlock';

const IconMapping = {
  lock: Lock,
  unlock: Unlock,
};

type IconArt = 'lock' | 'unlock' | string;

enum IconSize {
  XXXSmall = 2,
  XXSmall = 4,
  XSmall = 8,
  Small = 16,
  Medium = 32,
  Large = 64,
  XLarge = 128,
  XXLarge = 256,
  XXXLarge = 512,
}

interface IconProps {
  children: IconArt;
  size?: IconSize;
  color?: string;
  hover?: {icon: string; onHover: boolean};
}

const Icon = React.forwardRef((props: IconProps, ref: any) => {
  const {children = '', size = IconSize.Medium, color = 'black', hover = null, ...restProps} = props;
  return (
    <svg version="1.1" viewBox="0 0 512 512" width={size} height={size} ref={ref} {...restProps}>
      {!hover
        ? React.createElement(IconMapping[children], {color})
        : React.createElement(IconMapping[!hover.onHover ? children : hover.icon], {color})}
    </svg>
  );
});

Icon.displayName = 'Icon';

export {Icon};
