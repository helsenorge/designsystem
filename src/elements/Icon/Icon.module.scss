.icon {
  width: 4rem;
  height: 4rem;
  &--tiny {
    width: 1rem;
    height: 1rem;
  }
  &--small {
    width: 2rem;
    height: 2rem;
  }
  &--medium {
    width: 4rem;
    height: 4rem;
  }
  &--large {
    width: 8rem;
    height: 8rem;
  }
  &--big {
    width: 16rem;
    height: 16rem;
  }
  &--huge {
    width: 32rem;
    height: 32rem;
  }
}
