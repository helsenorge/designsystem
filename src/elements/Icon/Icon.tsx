import cn from 'classnames';
import React from 'react';
import {IconSizes} from '../../constants';
import styles from './Icon.module.scss';

interface IconProps extends React.HTMLProps<HTMLOrSVGImageElement> {
  children: string;
  size?: IconSizes;
}

function Icon(props: IconProps) {
  const Svg = require(`!svg-react-loader!../../icons/${props.children}.svg`);
  const classes = cn(
    styles.icon,
    {[styles['icon--tiny']]: props.size === IconSizes.Tiny},
    {[styles['icon--small']]: props.size === IconSizes.Small},
    {[styles['icon--medium']]: props.size === IconSizes.Medium},
    {[styles['icon--large']]: props.size === IconSizes.Large},
    {[styles['icon--big']]: props.size === IconSizes.Big},
    {[styles['icon--huge']]: props.size === IconSizes.Huge},
  );
  return <Svg className={classes} />;
}

Icon.defaultProps = {
  size: IconSizes.Medium,
};

export {Icon};
