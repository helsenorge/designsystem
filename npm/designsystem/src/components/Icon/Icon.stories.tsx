import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import Docs from '../../docs';
import Checkbox from '../Checkbox';
import additionalIconInformation from '../Icons/AdditionalIconInformation';
import { IconList, IconName } from '../Icons/IconNames';
import Search from '../Icons/Search';
import ExampleSvgIcon from '../Icons/Undo';
import Input from '../Input';
import LazyIcon from '../LazyIcon';
import Select from '../Select';
import Spacer from '../Spacer';

import Icon, { IconSize } from '.';

const meta = {
  title: '@helsenorge/designsystem-react/Components/Icon',
  component: Icon,
  parameters: {
    docs: {
      page: (): React.JSX.Element => <Docs component={Icon} />,
      description: {
        component: 'Icon lar deg vise et av flere ikoner i ulike størrelser og farger',
      },
    },
  },
  args: {
    color: 'black',
    hoverColor: 'black',
    isHovered: false,
  },
  argTypes: {
    color: {
      control: 'text',
    },
    hoverColor: {
      control: 'text',
    },
    isHovered: {
      control: 'boolean',
    },
  },
} satisfies Meta<typeof Icon>;

export default meta;

type Story = StoryObj<typeof Icon>;

export const Default: Story = {
  args: {
    svgIcon: ExampleSvgIcon,
  },
  render: args => {
    return (
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <div>
          <p>{'XLarge'}</p>
          <Icon {...args} size={IconSize.XLarge} />
        </div>
        <div>
          <p>{'Large'}</p>
          <Icon {...args} svgIcon={ExampleSvgIcon} size={IconSize.Large} />
        </div>
        <div>
          <p>{'Medium'}</p>
          <Icon {...args} svgIcon={ExampleSvgIcon} size={IconSize.Medium} />
        </div>
        <div>
          <p>{'Small'}</p>
          <Icon {...args} svgIcon={ExampleSvgIcon} size={IconSize.Small} />
        </div>
        <div>
          <p>{'XSmall'}</p>
          <Icon {...args} svgIcon={ExampleSvgIcon} size={IconSize.XSmall} />
        </div>
        <div>
          <p>{'XXSmall'}</p>
          <Icon {...args} svgIcon={ExampleSvgIcon} size={IconSize.XXSmall} />
        </div>
      </div>
    );
  },
};

export const Accessibility: Story = {
  args: {
    svgIcon: Search,
    size: IconSize.Small,
  },
  render: args => (
    <>
      <p>{'aria-label'}</p>
      <Icon {...args} ariaLabel="Search" />
      <Spacer size="4xs" />
      <p>{'ingen aria-label'}</p>
      <Icon {...args} />
    </>
  ),
};

type LazyIconStory = StoryObj<typeof LazyIcon>;

export const IconWall: LazyIconStory = {
  render: args => {
    const [searchText, setSearchText] = React.useState<string | undefined>();
    const [iconSize, setIconSize] = React.useState<IconSize>(IconSize.Medium);
    const [isHovered, setIsHovered] = React.useState<boolean>(false);

    const searchIcons = (searchText: string): string[] => {
      const lowerCaseSearchText = searchText.toLowerCase();

      return Object.keys(additionalIconInformation).filter(key => {
        const value = additionalIconInformation[key as keyof typeof additionalIconInformation];
        const matchKey = key.toLowerCase().includes(lowerCaseSearchText);
        const matchAlternativeName = value.alternativeName.toLowerCase().includes(lowerCaseSearchText);
        const matchCategories = value.categories.toLowerCase().includes(lowerCaseSearchText);

        return matchKey || matchAlternativeName || matchCategories;
      });
    };

    function includesCaseInsensitive(mainStr: string, searchStr: string): boolean {
      return mainStr.toLowerCase().includes(searchStr.toLowerCase());
    }

    const filterCamelCaseKeys = (searchText: string): string[] => {
      const lowercaseKeys = searchIcons(searchText).map(key => key.toLowerCase());
      return IconList.filter(camelCaseKey => lowercaseKeys.some(lowercaseKey => includesCaseInsensitive(camelCaseKey, lowercaseKey)));
    };

    const iconSizeKeys = Object.keys(IconSize).filter(key => isNaN(Number(key)));
    const filteredIcons = typeof searchText !== 'undefined' ? filterCamelCaseKeys(searchText) : IconList;

    return (
      <div>
        <div
          style={{
            background: '#fff',
            position: 'fixed',
            top: 0,
            left: 0,
            display: 'flex',
            alignItems: 'center',
            gap: '1rem',
            padding: '1rem',
            width: '100%',
          }}
        >
          <Input
            label={'Finn ikon'}
            onChange={e => {
              setSearchText(e.currentTarget.value);
            }}
          />
          <Select
            label={'Velg størrelse'}
            defaultValue={iconSizeKeys[3]}
            onChange={e => {
              setIconSize(parseInt(e.currentTarget.value, 10));
            }}
          >
            {iconSizeKeys.map((size, index) => {
              return <option key={index}>{size}</option>;
            })}
          </Select>
          <Checkbox
            label={'Hover'}
            onChange={() => {
              setIsHovered(!isHovered);
            }}
          />
        </div>
        <div
          style={{
            display: 'grid',
            gridTemplateColumns: '1fr 1fr 1fr',
            columnGap: '2rem',
            rowGap: '2rem',
            paddingTop: '10rem',
          }}
        >
          {filteredIcons.map((iconName, index) => {
            const additionalInfo = additionalIconInformation[iconName.toLowerCase() as keyof typeof additionalIconInformation];

            return (
              <div
                style={{ display: 'flex', width: 'fit-content', minWidth: '10rem', flexDirection: 'column', alignItems: 'center' }}
                key={index}
              >
                <LazyIcon {...args} iconName={iconName as IconName} size={iconSize} isHovered={isHovered} />
                <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}>
                  <span>{iconName}</span>
                  {additionalInfo && (
                    <>
                      <span style={{ fontSize: '1rem' }}>{'(' + additionalInfo.alternativeName + ')'}</span>
                      <Spacer />
                      <span style={{ fontSize: '1rem' }}>{additionalInfo.categories}</span>
                    </>
                  )}
                </div>
              </div>
            );
          })}
        </div>
      </div>
    );
  },
};
