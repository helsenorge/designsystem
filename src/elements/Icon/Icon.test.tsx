import React from 'react';
import {render} from 'react-testing-library';
import Icon from '.';

test('default variant renders correctly', () => {
  const {container} = render(<Icon>calendar</Icon>);
  expect(container).toMatchSnapshot();
});
