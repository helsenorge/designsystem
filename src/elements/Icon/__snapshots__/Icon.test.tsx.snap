// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default variant renders correctly 1`] = `
<div>
  <calendar.svg
    class="icon icon--medium"
  />
</div>
`;
